packages/data-source/src/DataSourceWidget.tsx

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2018-2020 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../style/index.css';
+
+import { ExpandableComponent } from '@elyra/ui-components';
+
+import { JupyterFrontEnd } from '@jupyterlab/application';
+import {
+  ReactWidget,
+  UseSignal,
+  Dialog,
+  showDialog
+} from '@jupyterlab/apputils';
+import { CodeCell, MarkdownCell } from '@jupyterlab/cells';
+import { CodeEditor } from '@jupyterlab/codeeditor';
+import { PathExt } from '@jupyterlab/coreutils';
+import { IDocumentManager } from '@jupyterlab/docmanager';
+import { DocumentWidget } from '@jupyterlab/docregistry';
+import { FileEditor } from '@jupyterlab/fileeditor';
+import { Notebook, NotebookPanel } from '@jupyterlab/notebook';
+import { IRenderMimeRegistry } from '@jupyterlab/rendermime';
+import { addIcon } from '@jupyterlab/ui-components';
+
+import { CommandRegistry } from '@lumino/commands';
+import { Message } from '@lumino/messaging';
+import { Signal } from '@lumino/signaling';
+import { Widget, Menu } from '@lumino/widgets';
+
+import React from 'react';
+
+import { DataSourceManager, IDataSource, ICodeTemplate } from './DataSource';
+
+/**
+ * The CSS class added to data source widget.
+ */
+const DATA_SOURCES_CLASS = 'elyra-DataSources';
+const DATA_SOURCES_HEADER_CLASS = 'elyra-dataSourcesHeader';
+const DATA_SOURCE_ITEM = 'elyra-dataSource-item';
+
+/**
+ * DataSourceDisplay props.
+ */
+interface IDataSourceDisplayProps {
+  dataSources: IDataSource[];
+  getCurrentWidget: () => Widget;
+  app: JupyterFrontEnd;
+}
+
+/**
+ * A React Component for data-sources display list.
+ */
+class DataSourceDisplay extends React.Component<IDataSourceDisplayProps> {
+  props: any;
+  commandRegistry: CommandRegistry;
+
+  constructor(props: any) {
+    super(props);
+  }
+
+  // Handle code snippet insert into an editor
+  private insertCodeSnippet = async (snippet: ICodeTemplate): Promise<void> => {
```

Comment:
I'm guessing most of this function is also used for actual code-snippets.  Are code snippets expected to implement `ICodeTemplate` or should data sources implement `ICodeSnippet`?  Somehow, the two should be refactored to share this code.

Seems like everything but `addSubMenu` and `renderDataSource` could be shared.